### DISCUSSION TITLE
Vertically centering text in a text box?

### DISCUSSION THREAD

**Discussion by amSnoop** [AvaloniaUI/Avalonia#10317] [Q&A]
I have it working just fine when any of the psuedoclasses are active, but when the box has text in it and is not focused, the text goes back to being top aligned.

From my Styles.axaml:

```xml
	<Style Selector="TextBox /template/ Border#PART_BorderElement">
		<Setter Property="CornerRadius" Value="0" />
		<Setter Property="VerticalAlignment" Value="Center" />
	</Style>
	<Style Selector="TextBox /template/ TextPresenter#PART_TextPresenter">
		<Setter Property="VerticalAlignment" Value="Center" />
	</Style>
	<Style Selector="TextBox:empty /template/ TextPresenter#PART_TextPresenter">
		<Setter Property="VerticalAlignment" Value="Center" />
	</Style>
	<Style Selector="TextBox:focus /template/ TextPresenter#PART_TextPresenter">
		<Setter Property="VerticalAlignment" Value="Center" />
	</Style>
	<Style Selector="TextBox:focus-within /template/ TextPresenter#PART_TextPresenter">
		<Setter Property="VerticalAlignment" Value="Center" />
	</Style>
	<Style Selector="TextBox:empty /template/ Border#PART_BorderElement">
		<Setter Property="BorderThickness" Value="1" />
		<Setter Property="Background" Value="{DynamicResource Alt}" />
		<Setter Property="BorderBrush" Value="{DynamicResource MoreHighlight}" />
		<Setter Property="Opacity" Value="1" />
	</Style>
	<Style Selector="TextBox:empty /template/ TextBlock#PART_Watermark">
		<Setter Property="Foreground" Value="White" />

		<!-- This does nothing -->
		<Setter Property="Opacity" Value="1" />
	</Style>
```

Behavior when empty:
![image](https://user-images.githubusercontent.com/49213176/218322317-bc1b4f9c-28c3-4768-80e4-99882b677569.png)
Behavior when focused:
![image](https://user-images.githubusercontent.com/49213176/218322339-7303c6b2-064a-4aeb-8757-c981c7ad52a1.png)
Behavior when unfocused:
![image](https://user-images.githubusercontent.com/49213176/218322358-30ff3788-0cba-4258-af72-5e8c4398fd1c.png)

From the DevTools:
![image](https://user-images.githubusercontent.com/49213176/218322697-477fe03c-108d-4e43-959d-6abeb568a0d8.png)
I am not sure why it shows that vertical alignment is being properly inherited from the custom style, but the real value is still "Stretch"

VerticalAlignment shows priority as a TemplatedParent, which I thought meant that if I overrode the template, the value would be overwritten.

I am using 0.10.18 on Windows 11.

The full files can be found [in this repository.](https://github.com/amSnoop/NewRenderWindow)

Thanks.

P.S. It looks like the Opacity of the watermark is set as a local value and therefore unable to be overwritten. Am I missing something with that or is that true? If it is true, is there a reason why it was made so it can't be overwritten?

**Comment by jaeaton** [comment]
Did you try VerticalContentAlignment="Center"

**Reply by timunie** [reply]
That's not the issue here. The question is, what is the "center" of a TextBlock? Is it the center of the bounding box or is it the center of using the base line of the text? This has an impact on how your eyes see the text. See for ecample: https://en.wikipedia.org/wiki/Baseline_(typography)

For example, take the picture below: Where do you think the text should be centered?
![different text lines](https://html.spec.whatwg.org/images/baselines.png)

**Reply by liufeias** [reply]
he means  the center of the bounding box
